aws-rds-dbcluster/src/main/java/software/amazon/rds/dbcluster/UpdateHandler.java

```diff
@@ -119,4 +139,95 @@ private boolean shouldRemoveFromGlobalCluster(
+                            isDBInstanceStabilized(proxyClient, dbClusterMember.dbInstanceIdentifier(), model))
+                    .handleError((createRequest, exception, client, model, callbackCtx) -> Commons.handleException(
+                            ProgressEvent.progress(model, callbackCtx),
+                            exception,
+                            DEFAULT_DB_CLUSTER_ERROR_RULE_SET
+                    ))
+                    .progress();
+
+            if (p.isFailed() || p.isInProgressCallbackDelay()) {
+                return p;
+            }
+            result = p;
+        }
+        return result;
+    }
+
+    private boolean shouldRebootCluster(final ResourceModel previousResourceState, final ResourceModel desiredResourceState, boolean isRollback) {
+        return !isRollback &&
+                !Objects.equals(previousResourceState.getEngineVersion(), desiredResourceState.getEngineVersion()) &&
+                !Objects.equals(previousResourceState.getDBInstanceParameterGroupName(), desiredResourceState.getDBInstanceParameterGroupName());
+    }
+
+    private boolean isDBInstanceStabilized(final ProxyClient<RdsClient> client, final String dbInstanceIdentifier, final ResourceModel resourceModel) {
+        final DBInstance dbInstance = fetchDBInstance(client, dbInstanceIdentifier);
+        final DBCluster dbCluster = fetchDBCluster(client, resourceModel);
+
+        return "available".equals(dbInstance.dbInstanceStatus()) &&
+                isParameterGroupApplied(dbInstance) &&
+                isClusterParameterGroupApplied(dbCluster, dbInstance);
+    }
+    private boolean isParameterGroupApplied(final DBInstance dbInstance) {
+        return Optional.ofNullable(dbInstance.dbParameterGroups()).orElse(Collections.emptyList()).stream()
+                .allMatch(pg -> IN_SYNC_STATUS.equals(pg.parameterApplyStatus()));
+    }
+
+    private boolean isClusterParameterGroupApplied(final DBCluster dbCluster, final DBInstance dbInstance) {
+        final DBClusterMember member = dbCluster.dbClusterMembers().stream()
+                .filter(m -> dbInstance.dbInstanceIdentifier().equalsIgnoreCase(m.dbInstanceIdentifier()))
+                .findFirst().get();
+        return IN_SYNC_STATUS.equals(member.dbClusterParameterGroupStatus());
+    }
+
+    protected boolean withProbing(
+            final CallbackContext context,
+            final String probeName,
+            final int nProbes,
+            final Supplier<Boolean> checker
+    ) {
+        final boolean check = checker.get();
+        if (!config.isProbingEnabled()) {
+            return check;
+        }
+        if (!check) {
+            context.flushProbes(probeName);
+            return false;
+        }
+        context.incProbes(probeName);
+        if (context.getProbes(probeName) >= nProbes) {
+            context.flushProbes(probeName);
+            return true;
+        }
+        return false;
+    }
```

Comment:
As this is a travelling bit of code, let's move it to BaseHandler: contextually it is a generic helper

Reply:
TBF, we probably want to move a lot of the code in this PR to a common place. E.g. there is a lot of duplicated code between DBInstance and DBCluster. I think we want to do that as a separate PR.

Reply:
I think we might want to address the underlying reason for having this bit of code in the first place, but that is a bigger project.